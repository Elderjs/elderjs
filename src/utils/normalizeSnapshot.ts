import windowsPathFix from './windowsPathFix';

const normalizeSnapshot = (val) => {
  if (Object.prototype.toString.call(val) === '[object String]') {
    return windowsPathFix(val);
  }
  if (Object.prototype.toString.call(val) === '[object Array]') {
    return val.map(normalizeSnapshot);
  }
  if (Object.prototype.toString.call(val) === '[object Object]') {
    return Object.keys(val).reduce((out, cv) => {
      // eslint-disable-next-line no-param-reassign
      out[cv] = normalizeSnapshot(val[cv]);
      return out;
    }, {});
  }
  return val;
};

export default normalizeSnapshot;
