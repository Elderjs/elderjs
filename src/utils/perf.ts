import { performance, PerformanceObserver } from 'perf_hooks';
import { Elder } from '..';
import Page from './Page';
/**
 * A little helper around perf_hooks.
 * Returns an object with a start and end function.
 *
 * This allows you to pass in a page.perf.start('name') and then page.perf.end('name') and the result is stored in a timings array.
 *
 */

export type TPerfTiming = { name: string; duration: number };

export type TPerfTimings = TPerfTiming[];

const perf = (page: Page | Elder, force = false) => {
  if (page.settings.debug.performance || force) {
    let obs = new PerformanceObserver((items) => {
      items.getEntries().forEach((entry) => {
        if (entry.name.includes(page.uid)) {
          page.perf.timings.push({ name: entry.name.replace(`-${page.uid}`, ''), duration: entry.duration });
          performance.clearMarks(entry.name);
        }
      });
    });

    // eslint-disable-next-line no-param-reassign
    page.perf = {
      timings: [],
      /**
       * Marks the performance timeline with {label}-start.
       *
       * @param {String} label
       */
      start: (label: string) => {
        performance.mark(`${label}-start-${page.uid}`);
      },
      /**
       * Marks the performance timeline with {label}-stop
       * It then records triggers a recording of the measurement.
       * @param {*} label
       */
      end: (label) => {
        performance.mark(`${label}-end-${page.uid}`);
        performance.measure(`${label}-${page.uid}`, `${label}-start-${page.uid}`, `${label}-end-${page.uid}`);
      },
      stop: () => {
        if (obs) obs.disconnect();
        obs = null;
      },
    };

    obs.observe({ entryTypes: ['measure'] });
  } else {
    const placeholder = () => {};
    // eslint-disable-next-line no-param-reassign
    page.perf = {
      timings: [],
      start: placeholder,
      end: placeholder,
      stop: placeholder,
    };
  }

  // eslint-disable-next-line no-param-reassign
  page.perf.prefix = (pre) => {
    return { start: (name) => page.perf.start(`${pre}.${name}`), end: (name) => page.perf.end(`${pre}.${name}`) };
  };
};

export default perf;

export const displayPerfTimings = (timings: TPerfTimings) => {
  const display = timings.sort((a, b) => a.duration - b.duration).map((t) => ({ ...t, ms: t.duration }));
  console.table(display, ['name', 'ms']);
};
