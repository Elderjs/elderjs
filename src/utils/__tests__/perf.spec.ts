import normalizeSnapshot from '../normalizeSnapshot.js';
import perf from '../perf';
import { describe, it, expect, vi, beforeAll, beforeEach } from 'vitest';

beforeAll(() => {
  vi.resetModules();
});

beforeEach(() => {
  vi.resetModules();
});

describe('#perf', () => {
  it('works in performance and mocks', () => {
    function MockPage() {
      this.uid = 'xxxxxxxx';
      this.htmlString = '';
      this.settings = {
        debug: {
          performance: true,
        },
      };
    }

    const mockPage = new MockPage();

    // mutate
    perf(mockPage);

    mockPage.perf.start('test');
    mockPage.perf.end('test');

    const prefixed = mockPage.perf.prefix('prefix');

    prefixed.start('prefix');
    prefixed.end('prefix');

    mockPage.perf.stop();

    expect(normalizeSnapshot(mockPage.perf.timings[0].name)).toBe('test');
    expect(normalizeSnapshot(mockPage.perf.timings[1].name)).toBe('prefix.prefix');
  });

  it('works in non performance', () => {
    function MockPage() {
      this.uid = 'xxxxxxxx';
      this.htmlString = '';
      this.settings = {
        debug: {
          performance: false,
        },
      };
    }

    const mockPage = new MockPage();

    // mutate
    perf(mockPage);

    mockPage.perf.start('test');
    mockPage.perf.end('test');
    mockPage.perf.stop();

    expect(mockPage.perf.timings).toEqual([]);
  });
});
