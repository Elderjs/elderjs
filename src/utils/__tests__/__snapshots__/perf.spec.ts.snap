// Vitest Snapshot v1

exports[`#perf > works in performance and mocks 1`] = `
{
  "htmlString": "",
  "perf": {
    "end": [Function],
    "prefix": [Function],
    "start": [Function],
    "stop": [Function],
    "timings": [
      {
        "duration": 0.026834,
        "name": "test",
      },
      {
        "duration": 0.014667,
        "name": "prefix.prefix",
      },
    ],
  },
  "settings": {
    "debug": {
      "performance": true,
    },
  },
  "uid": "xxxxxxxx",
}
`;

exports[`#perf works in performance and mocks 1`] = `
Object {
  "htmlString": "",
  "perf": Object {
    "end": [Function],
    "prefix": [Function],
    "start": [Function],
    "stop": [Function],
    "timings": Array [
      Object {
        "duration": 0.05,
        "name": "Page",
      },
    ],
  },
  "settings": Object {
    "debug": Object {
      "performance": true,
    },
  },
  "uid": "xxxxxxxx",
}
`;
